File: GO/a2zSheet/step17_tries/longest_string_with_all_prefixes.cpp
#include <bits/stdc++.h> 

struct Node{
    Node* links[26];
    bool flag = false;
};

class Trie{
    Node* root;
    public:
    Trie(){
        root = new Node();
    }
    void insert(string word){
        Node* temp = root;
        for(auto &c:word){
            if(!temp->links[c-'a']){
                temp->links[c-'a'] = new Node();
            }
            temp = temp->links[c-'a'];
        }

        temp->flag = true;
    }

    int countPrefixs(string word){
        Node* temp = root;
        int cnt = 0;
        for(auto &c:word){
            if(temp->links[c-'a']){
                temp = temp->links[c-'a'];
                cnt += temp->flag == 1;
            }
            else return 0;
        }

        return cnt;
    }
};

string completeString(int n, vector<string> &a){

    Trie t;
    for(auto &c:a){
        t.insert(c);
    }

    string ans = "";
    int gcnt = 0;
    for(auto &c:a){
        int cnt = t.countPrefixs(c);
        if(cnt!=c.size())continue;
        if(cnt>gcnt){
            ans = c;
            gcnt = cnt;
        }
        else if(cnt==gcnt){
            if(c<ans)ans = c;
        }
    }
    if(ans=="")return "None";
    return ans;
}

File: GO/a2zSheet/step17_tries/longest_string_with_all_prefixes.go
package main

import (
	"fmt"
)

type Node struct {
	links  [26]*Node
	isLeaf bool
}

type Trie struct {
	root *Node
}

func Constructor() Trie {
	return Trie{root: &Node{}}
}

func (t *Trie) Insert(word string) {
	temp := t.root
	for _, c := range word {
		index := c - 'a'
		if temp.links[index] == nil {
			temp.links[index] = &Node{}
		}
		temp = temp.links[index]
	}
	temp.isLeaf = true
}

func (t *Trie) CountPrefixs(word string) int {
	temp := t.root
	cnt := 0
	for _, c := range word {
		if temp.links[c-'a'] != nil {
			temp = temp.links[c-'a']
			if temp.isLeaf {
				cnt++
			}
		} else {
			return 0
		}
	}
	return cnt
}

func completeString(n int, a []string) string {
	t := Constructor()
	for _, str := range a {
		t.Insert(str)
	}

	ans := ""
	gcnt := 0
	for _, str := range a {
		cnt := t.CountPrefixs(str)
		if cnt != len(str) {
			continue
		}
		if cnt > gcnt {
			ans = str
			gcnt = cnt
		} else if cnt == gcnt {
			if str < ans {
				ans = str
			}
		}
	}

	if ans == "" {
		return "None"
	}
	return ans
}

func main() {
	a := []string{"n", "ni", "nin", "ninj", "ninja", "ninga"}
	n := len(a)
	fmt.Println(completeString(n, a))
}
